### PR TITLE
Nc news 11

### PR DESCRIPTION
**Pull request by charli353**

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -182,6 +183,75 @@ describe("CORE: GET - /api/articles/:article_id", () => {
...
+          .get("/api/articles?topic=mitch")
+          .expect(200)
+          .then(({ body }) => {
+            expect(body.articles.length).not.toBe(0)
+            body.articles.forEach((article) => {
+                expect(article).toHaveProperty("topic", ('mitch'))
+            })        
+          });
+      });
+      test("200: Invalid topic query within endpoint returns empty array", () => {
```
Comment (Keb84) [inline __tests__/app.test.js]: Here you have described this as an invalid topic however if something is invalid would we expect at status code 200 🤔

File: __tests__/app.test.js [hunk 2]
```diff
@@ -182,6 +183,75 @@ describe("CORE: GET - /api/articles/:article_id", () => {
...
+        return request(app)
+          .get("/api/articles?topic=faketopic")
+          .expect(200)
+          .then(({ body }) => {
+            expect(body).toEqual({articles: []})
```
Comment (Keb84) [inline __tests__/app.test.js]: If 'faketopic' isn't in the database would we really be getting an empty array here 👀 . Maybe some kind of error message would be more appropriate 💭

File: __tests__/app.test.js [hunk 3]
```diff
@@ -182,6 +183,75 @@ describe("CORE: GET - /api/articles/:article_id", () => {
...
+          .then(({ body }) => {
+          
+            expect(body.articles.length).not.toBe(0)
+            const ids = body.articles.map((id) => {
+              return id.article_id
+            })        
+            expect(ids).toBeSorted({ descending: true })
+          })
+        })
+        test("200: Invalid sort by query within endpoint returns default sort (DESC BY DATE)", () => {
```
Comment (Keb84) [inline __tests__/app.test.js]: Again here your description mentions that this query is invalid so would you expect a status 200 🤔

File: __tests__/app.test.js [hunk 4]
```diff
@@ -182,6 +183,75 @@ describe("CORE: GET - /api/articles/:article_id", () => {
...
+          return request(app)
+            .get("/api/articles?sort_by=not_a_column")
+            .expect(200)
+            .then(({ body }) => {
+      
+              expect(body.articles.length).not.toBe(0)
+              const dates = body.articles.map((article) => {
+                return article.created_at
+              })        
+              expect(dates).toBeSorted({ descending: true })
```
Comment (Keb84) [inline __tests__/app.test.js]: If we pass a sort_by that is not valid maybe an error status would be more appropriate?

File: __tests__/app.test.js [hunk 5]
```diff
@@ -182,6 +183,75 @@ describe("CORE: GET - /api/articles/:article_id", () => {
...
+            .expect(200)
+            .then(({ body }) => {
+              expect(body.articles.length).not.toBe(0)
+              const ids = body.articles.map((id) => {
+                return id.article_id
+              })        
+              expect(ids).toBeSorted({ ascending: true })
+            });
+        });
+        test("200: Invalid Order query returns articles in default order", () => {
```
Comment (Keb84) [inline __tests__/app.test.js]: Again your description would need refactoring - same as before

File: __tests__/app.test.js [hunk 6]
```diff
@@ -182,6 +183,75 @@ describe("CORE: GET - /api/articles/:article_id", () => {
...
+          return request(app)
+            .get("/api/articles?sort_by=article_id&order=FAKEORDER")
+            .expect(200)
+            .then(({ body }) => {
+              expect(body.articles.length).not.toBe(0)
+              const ids = body.articles.map((id) => {
+                return id.article_id
+              })        
+              expect(ids).toBeSorted({ descending: true })
```
Comment (Keb84) [inline __tests__/app.test.js]: Same point as before - as this is an invalid order query you should be expecting a bad request

File: __tests__/app.test.js [hunk 7]
```diff
@@ -182,6 +183,75 @@ describe("CORE: GET - /api/articles/:article_id", () => {
...
+            });
+        });
```
Comment (Keb84) [inline __tests__/app.test.js:259]: An addition test for is the topic is valid but there are no articles would be a good test to check here 🧐

File: db/error-handler.js [hunk 8]
```diff
@@ -19,6 +18,9 @@ function errorHandler(err, req, res, next) {
     else if (err === 'Error: 404 - Not Found'){
         res.status(404).send({ Error: 'ID Does Not Exist' })
     }
+    else if (err === 'Error: 200 - No Articles'){
+        res.status(200).send({articles: []})
+    }
```
Comment (Keb84) [inline db/error-handler.js:33]: Do you really want a error with a 200 status code? Would that be what you would expect from an error? 🧐

File: db/models/articles.model.js [hunk 9]
```diff
@@ -39,14 +22,62 @@ function retrieveRelevantComments(id) {
...
         LEFT JOIN comments ON a.article_id = comments.article_id
-        GROUP BY a.article_id ORDER BY a.created_at DESC;`)
-        .then(({rows}) => {
-            return rows
-        })
+        GROUP BY a.article_id ORDER BY a.${query.sort_by} ${query.order};`)
+            .then(({rows}) => {
+                return queryCheck(query, rows, key)
+            })
+    }
```
Comment (Keb84) [inline db/models/articles.model.js:68]: This model is very big. Your query strings are very similar. Could this be refactored to be more DRY? 💭

**Review by Keb84 — CHANGES_REQUESTED** [review]
Nice work on this ticket 🤸 please work on the comments made and resubmit the ticket when you're done 🎸

**Review by Dutton1990 — COMMENTED** [review]
Nice job with the refactoring! 

I can't see a test for a valid topic with no articles - there's a topic (paper) that exists but has no articles. In that case you can return an empty [] and 200. 

You don't need to resubmit another PR now though, you are free to merge whether you want to implement the above now or at another point!

This ticket has a lot to check for, so well done!

**Pull request merged**